frame/contracts/src/benchmarking/mod.rs

```diff
@@ -893,15 +892,80 @@ benchmarks! {
 			}],
 			call_body: Some(body::repeated(r * API_BENCHMARK_BATCH_SIZE, &[
 				Instruction::I32Const(0), // value_ptr
-				Instruction::I32Const(max_bytes as i32), // value_len
+				Instruction::I32Const(0), // value_len
 				Instruction::Call(0),
 				Instruction::Drop,
 			])),
 			.. Default::default()
 		});
 		let instance = Contract::<T>::new(code, vec![])?;
-		let origin = RawOrigin::Signed(instance.caller.clone());
-	}: call(origin, instance.addr, 0u32.into(), Weight::MAX, None, vec![])
+	}: {
+		<Contracts<T>>::bare_call(
+			instance.caller,
+			instance.account_id,
+			0u32.into(),
+			Weight::MAX,
+			None,
+			vec![],
+			true,
+			Determinism::Deterministic,
+		)
+		.result?;
+	}
+
+	seal_debug_message_per_kb {
+		// Vary size of input in kilobytes up to maximum allowed contract memory.
+		let i in 0 .. (T::Schedule::get().limits.memory_pages * 64);
+		// All ASCII codes + an invalid utf-8 code: 128
+		let codes = (0..129).collect::<Vec<_>>();
+		// Largest debug message we can benchmark for is limited by maximum allowed
+		// contract memory which is defined by T::Schedule::get().limits.memory_pages,
+		// which to the moment is 1 Mb.
+		// We can't however initialize every byte of it, as it would bloat wasm module size which is
+		// limited by T::MaxCodeLen::get(), being far less than 1Mb.
```

Comment:
You should only let `i` run up until the amount of memory you are able to initialise. Otherwise the vast amount of sizes are against invalid sequence `0`. This will not copy it to the buffer and takes less time. Hence we underestimate. In benchmarks we should always benchmark the worst case. Hence only initialise with valid characters which will be copied to the buffer and incur the biggest cost.

Just let `i` run up to the allowed buffer size capped by the max code size. We probably can't use batching here and just do one run (otherwise the buffer might fill up).

Reply:
> Are you sure? I think benchmarks skip the size and other validation steps.

Yep, it [fails with `Original code too large`](https://github.com/paritytech/substrate/blob/ag-debug-buffer/frame/contracts/src/wasm/prepare.rs#L566) when preparing the wasm module (only when it's really too large ofc)

> You should only let `i` run up until the amount of memory you are able to initialise. Otherwise the vast amount of sizes are against invalid sequence `0`. This will not copy it to the buffer and takes less time. Hence we underestimate. In benchmarks we should always benchmark the worst case. Hence only initialise with valid characters which will be copied to the buffer and incur the biggest cost.
> 
> We probably can't use batching here because the buffer will be full in the later runs.


Agreed. Updated the benchmark.